src/software/ai/navigator/obstacle/new_obstacle/obstacle.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <sstream>
+
+#include "shared/constants.h"
+#include "software/new_geom/point.h"
+#include "software/new_geom/polygon.h"
+#include "software/new_geom/segment.h"
+#include "software/new_geom/util/distance.h"
+#include "software/new_geom/util/intersects.h"
+
+/**
+ * An obstacle is an area to avoid for navigation
+ */
+class Obstacle
+{
+   public:
+    virtual ~Obstacle() = default;
+
+    /**
+     * Determines whether the given Point is contained within this Obstacle
+     *
+     * @return whether the Point p is contained within this Obstacle
+     */
+    virtual bool contains(const Point& p) const = 0;
+
+    /**
+     * Gets the minimum distance from the obstacle to the point
+     *
+     * @param point Point to get distance to
+     *
+     * @return distance to point
+     */
+    virtual double minDistance(const Point& p) const = 0;
+
+    /**
+     * Determines whether the given Segment intersects this Obstacle
+     *
+     * @return true if the given Segment intersects this Obstacle
+     */
+    virtual bool intersects(const Segment& segment) const = 0;
+
+    /**
+     * Output string to describe the obstacle
+     *
+     * @return string that describes the obstacle
+     */
+    virtual std::string toString(void) const = 0;
+};
+
+/**
+ * We use a pointer to obstacles to support inheritance
+ */
+using ObstaclePtr = std::shared_ptr<Obstacle>;
```

Comment:
Sorry, I'm not quite following why this is required?

Reply:
the idea is that we need to use pointers to the virtual class to use inheritance so ObstaclePtr is a shortcut

Reply:
Ahhh I see. Perhaps we could change this to something like `convenience typedef` instead? I was just a bit confused with the current phrasing because I tried to read too much into it.

Reply:
fixed it